include/nbl/builtin/hlsl/prefix_sum_blur.hlsl

```diff
@@ -0,0 +1,262 @@
+    typename Sampler,
+    uint16_t WorkgroupSize,
+    class device_capabilities=void> // TODO: define concepts for the Box1D and apply constraints
+struct Box1D
+{
+    // TODO: Generalize later on when Francesco enforces accessor-concepts in `workgroup` and adds a `SharedMemoryAccessor` concept
+    struct ScanSharedAccessorWrapper
+    {
+        void get(const uint16_t ix, NBL_REF_ARG(float32_t) val)
+        {
+            val = base.template get<float32_t, uint16_t>(ix);
+        }
+
+        void set(const uint16_t ix, const float32_t val)
+        {
+            base.template set<float32_t, uint16_t>(ix, val);
+        }
+
+        void workgroupExecutionAndMemoryBarrier()
+        {
+            base.workgroupExecutionAndMemoryBarrier();
+        }
+
+        ScanSharedAccessor base;
+    };
+
+    void operator()(
+        NBL_REF_ARG(DataAccessor) data,
+        NBL_REF_ARG(SharedAccessor) scratch,
+        NBL_REF_ARG(ScanSharedAccessor) scanScratch,
+        NBL_REF_ARG(Sampler) boxSampler,
+        const uint16_t channel)
+    {
+        const uint16_t end = data.linearSize();
+        const uint16_t localInvocationIndex = workgroup::SubgroupContiguousIndex();
+
+        // prefix sum
+        // note the dynamically uniform loop condition 
+        for (uint16_t baseIx = 0; baseIx < end;)
+        {
+            const uint16_t ix = localInvocationIndex + baseIx;
+            float32_t input = data.template get<float32_t>(channel, ix);
+            // dynamically uniform condition
+            if (baseIx != 0)
+            {
+                // take result of previous prefix sum and add it to first element here
+                if (localInvocationIndex == 0)
+                    input += scratch.template get<float32_t>(baseIx - 1);
+            }
+            // need to copy-in / copy-out the accessor cause no references in HLSL - yay!
+            ScanSharedAccessorWrapper scanScratchWrapper;
+            scanScratchWrapper.base = scanScratch;
+            const float32_t sum = workgroup::inclusive_scan<plus<float32_t>, WorkgroupSize, device_capabilities>::template __call(input, scanScratchWrapper);
+            scanScratch = scanScratchWrapper.base;
+            // loop increment
+            baseIx += WorkgroupSize;
+            // if doing the last prefix sum, we need to barrier to stop aliasing of temporary scratch for `inclusive_scan` and our scanline
+            // TODO: might be worth adding a non-aliased mode as NSight says nr 1 hotspot is barrier waiting in this code
+            if (end + ScanSharedAccessor::Size > SharedAccessor::Size)
+                scratch.workgroupExecutionAndMemoryBarrier();
+            // save prefix sum results
+            if (ix < end)
+                scratch.template set<float32_t>(ix, sum);
+            // previous prefix sum must have finished before we ask for results
+            scratch.workgroupExecutionAndMemoryBarrier();
+        }
+
+        const float32_t last = end - 1;

```

Comment:
I'd try splitting the Box1D (or what I'd call `Blur1D`) into two separate functors:
1. multi-wg-wide prefix sum
2. the SAT sampling

don't do it but leave a comment about it